analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/RestrictDeserializedTypesBase.cs

```diff
@@ -26,4 +29,169 @@ public abstract class RestrictDeserializedTypesBase : SymbolicRuleCheck
+            && !ThrowsOrReturnsNull(declaration))
+        {
+            resolveTypeDeclaration = declaration;
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    private static bool UnsafeLosFormatter(ProgramState state, IObjectCreationOperationWrapper objectCreation) =>
+        objectCreation.Type.Is(LosFormatter)
+        && objectCreation.ArgumentValue("enableMac") switch
+        {
+            null => true,
+            { } enableMacArgument => state[enableMacArgument]?.HasConstraint(BoolConstraint.True) is not true
+        };
+
+    private ProgramState ProcessBinderAssignment(ProgramState state, IAssignmentOperationWrapper assignment)
+    {
+        if (BinderAssignmentInstance(state, assignment) is { } instance)
+        {
+            var constraint = BinderIsSafe(state, assignment, out var bindToTypeDeclaration)
+                ? SerializationConstraint.Safe
+                : SerializationConstraint.Unsafe;
+            if (constraint == SerializationConstraint.Unsafe)
+            {
+                additionalLocationsForOperations[instance] = bindToTypeDeclaration;
+            }
+            state = state.SetOperationConstraint(instance, constraint);
+
+            if (instance.TrackedSymbol() is { } symbol)
+            {
+                if (constraint == SerializationConstraint.Unsafe)
+                {
+                    additionalLocationsForSymbols[symbol] = bindToTypeDeclaration;
+                }
+                state = state.SetSymbolConstraint(symbol, constraint);
+            }
+            return state;
+        }
+        return null;
+    }
+
+    private static IOperation BinderAssignmentInstance(ProgramState state, IAssignmentOperationWrapper assignment) =>
+        state.ResolveCaptureAndUnwrapConversion(assignment.Target).AsPropertyReference() is { Property.Name: "Binder", Instance: { } propertyInstance }
+        && propertyInstance.Type.IsAny(FormattersWithBinder)
+            ? state.ResolveCaptureAndUnwrapConversion(propertyInstance)
+            : null;
+
+    private bool BinderIsSafe(ProgramState state, IAssignmentOperationWrapper assignment, out SyntaxNode bindToTypeDeclaration)
+    {
+        if (state[assignment.Value]?.HasConstraint(ObjectConstraint.Null) is true)
+        {
+            bindToTypeDeclaration = null;
+            return false;
+        }
+        else
+        {
+            bindToTypeDeclaration = DeclarationCandidates(state.ResolveCaptureAndUnwrapConversion(assignment.Value)).FirstOrDefault(IsBindToType);
+            return bindToTypeDeclaration is null || ThrowsOrReturnsNull(bindToTypeDeclaration);
+        }
+    }
+
+    private IEnumerable<SyntaxNode> DeclarationCandidates(IOperation operation) =>
+        SemanticModel.GetTypeInfo(operation.Syntax).Type?.DeclaringSyntaxReferences.SelectMany(x => x.GetSyntax().ChildNodes());
```

Comment:
```suggestion
        operation.Type?.DeclaringSyntaxReferences.SelectMany(x => x.GetSyntax().ChildNodes());
```

Reply:
This breaks a bunch of tests.

Reply:
I forgot to mention: also change the `UnsafeResolver` method. Unwrap the operation before passing it to the `DeclarationCadidates` method:

```cs
else if (DeclarationCandidates(state.ResolveCaptureAndUnwrapConversion(operation))?.FirstOrDefault(IsResolveType) is { } declaration
```

Then the UTs will remain green.